ios/MaasCore/Sources/MaasCore/Api/Utils/RefreshToken+Publishers.swift

```diff
@@ -0,0 +1,35 @@
+import Combine
+
+private var isRefreshing = false
+private var activeRefreshTokenPublisher: AnyPublisher<Void, Never>!
+static var access = NSLock()
+
+extension ApiConfig {
+        
+    func refreshTokenPublisher() -> AnyPublisher<Void, Never> {
+        Deferred { [unowned self] () -> AnyPublisher<Void, Never> in
+            
+            access.lock(); defer { access.unlock() }
+            
+            if !isRefreshing {
+                isRefreshing = true
+                activeRefreshTokenPublisher = self.performTokenRefresh()
+                    .handleEvents(receiveCompletion: { _ in isRefreshing = false })
```

Comment:
Need to think through whether this assignment should also be guarded, e.g. in a concurrent scenario, what happens when
1. (thread 1) enters Deferred block, if !isRefreshing returns false, i.e. `isRefreshing == true`
2. (thread 2) `activeRefreshTokenPublisher` completes and fires `receiveCompletion`, setting isRefreshing to false
3. (thread 1) returns (the already completed) `activeRefreshTokenPublisher`

Will a new subscriber to `activeRefreshTokenPublisher` after it has already completed get an emission?

Reply:
`activeRefreshTokenPublisher ` is constructed from a `Future`, all subsequent subscriptions to a publisher, that is constructed from a future will receive the same value, and finish events.
example shows 2️⃣ subscription that subscribes after `Future` is completed, receives the same value immediately, without any delay.
 
https://user-images.githubusercontent.com/70565612/108203286-f2bd7600-712a-11eb-9090-ed2974c3d18b.mp4

Reply:
Just to keep it more clear - moved the `isRefreshing = false` to token refresh completion handling with the same NSLog protection